core/lib/Cardano/Address/KeyHash.hs

```diff
@@ -0,0 +1,313 @@
+    deriving (Generic, Show, Ord, Eq)
+instance NFData KeyHash
+
+-- | Construct an 'KeyHash' from raw 'ByteString' (28 bytes).
+--
+-- @since 3.0.0
+keyHashFromBytes :: (KeyRole, ByteString) -> Maybe KeyHash
+keyHashFromBytes (cred, bytes)
+    | BS.length bytes /= credentialHashSize = Nothing
+    | otherwise = Just $ KeyHash cred bytes
+
+-- | Encode a 'KeyHash' to bech32 'Text' or hex is key role unknown.
+--  If one wants to include, valid in governance roles only, additional byte
+--  as specified in CIP-0129, the function needs to be called with withByte=true.
+--
+-- @since 3.0.0
+keyHashToText :: KeyHash -> Maybe GovernanceType -> Text
+keyHashToText (KeyHash cred keyHash) govType = case cred of
+    PaymentShared ->
+        T.decodeUtf8 $ encode (EBech32 CIP5.addr_shared_vkh) keyHash
+    DelegationShared ->
+        T.decodeUtf8 $ encode (EBech32 CIP5.stake_shared_vkh) keyHash
+    Payment ->
+        T.decodeUtf8 $ encode (EBech32 CIP5.addr_vkh) keyHash
+    Delegation ->
+        T.decodeUtf8 $ encode (EBech32 CIP5.stake_vkh) keyHash
+    Policy ->
+        T.decodeUtf8 $ encode (EBech32 CIP5.policy_vkh) keyHash
+    Representative -> case govType of
+        Just CIP0105 ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.drep_vkh) keyHash
+        _ ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.drep) $ keyHashAppendByteCIP0129 keyHash cred
+    CommitteeCold -> case govType of
+        Just CIP0105 ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_cold_vkh) keyHash
+        _ ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_cold) $ keyHashAppendByteCIP0129 keyHash cred
+    CommitteeHot -> case govType of
+        Just CIP0105 ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_hot_vkh) keyHash
+        _ ->
+            T.decodeUtf8 $ encode (EBech32 CIP5.cc_hot) $ keyHashAppendByteCIP0129 keyHash cred
+    Unknown ->
+        T.decodeUtf8 $ encode EBase16 keyHash
+
+-- | In accordance to CIP-0129 (https://github.com/cardano-foundation/CIPs/tree/master/CIP-0129)
+--   one byte is prepended to vkh only in governance context. The rules how to contruct it are summarized
+--   below
+--
+--   drep       0010....
+--   hot        0000....    key type
+--   cold       0001....
+--
+--   keyhash    ....0010
+--   This is on top of X_vkh, where X={drep, cc_hot, cc_hot}, which lacks the additional byte.
+--   In `keyHashFromText` we additionally
+--   support reading legacy X which also lacks the additional byte, and has the same payload as
+--   as the corresponding X_vkh.
+keyHashAppendByteCIP0129 :: ByteString -> KeyRole -> ByteString
+keyHashAppendByteCIP0129 payload cred =
+    maybe payload (`BS.cons` payload) bytePrefix
+  where
+    bytePrefix = case cred of
+        Representative -> Just 0b00100010
+        CommitteeCold -> Just 0b00010010
+        CommitteeHot -> Just 0b00000010
+        _ -> Nothing
+
+-- | Construct a 'KeyHash' from 'Text'. It should be
+-- Bech32 encoded text with one of following hrp:
```

Comment:
```suggestion
-- Bech32 encoded text with one of following hrp:

```

Need a newline for the list to be properly formatted in Haddock

Reply:
indeed, done